tools/toitlsp/lsp/project_uri.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (C) 2023 Toitware ApS.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; version
+// 2.1 only.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// The license can be found in the file `LICENSE` in the top level
+// directory of this repository.
+
+package lsp
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/sourcegraph/go-lsp"
+	"github.com/toitware/toit.git/toitlsp/lsp/uri"
+)
+
+func computeProjectURI(documentUri lsp.DocumentURI) (lsp.DocumentURI, error) {
+	path := uri.URIToPath(documentUri)
+	segments := strings.Split(filepath.ToSlash(path), "/")
+	// Find the first '.packages' segment. We assume that the project root is
+	// the parent of this segment.
+	for i := 0; i < len(segments); i++ {
+		if segments[i] == ".packages" {
```

Comment:
I wonder if we should have called this something more Toit-specific.

Reply:
Same for `package.yaml` and `package.lock`.

We should maybe try to rename them...